lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -913,112 +912,182 @@ def test_get_inventory(
+                0,
+                b"",
+                1,
+                1,
+                b"mock-tar: /mock/result.tar.xz: Cannot open: No such mock-file or mock-directory",
+            ),
+            # Unexpected failure
+            ("/usr/bin/tar", False, 0, b"", 1, 1, b"mock-tar: bolt out of the blue!"),
+            # Hang, never returning output nor an exit code
+            ("/usr/bin/tar", False, 0, b"", None, None, b""),
+        ),
+    )
+    def test_tarfile_extract(
+        self,
+        monkeypatch,
+        tmp_path,
+        tar_path: str,
+        popen_fail: bool,
+        wait_cnt: int,
+        peek_return: Optional[bytes],
+        poll_return: Optional[int],
+        proc_return: int,
+        stderr_contents: Optional[bytes],
+    ):
+        """Test to check Tarball.extract behaviors"""
         tar = Path("/mock/result.tar.xz")
-        path = Path("subdir/f11.txt")
-
-        class MockTarFile:
-            def extractfile(self, path):
-                return None
-
-        def fake_tarfile_open(self, path):
-            return MockTarFile()
+        path = "metadata.log"
+        stdout_contents = b"[test]\nfoo=bar\n"
+
+        class MockBufferedReader(io.BufferedReader):
+            def __init__(self, contents: bytes):
+                # No effect, other than to quiet the linter
+                None if True else super().__init__(io.RawIOBase())
+                self.contents = contents
+                self.loop_count = wait_cnt
+
+            def close(self) -> None:
+                raise AssertionError(
+                    "This test doesn't expect the stream to be closed."
+                )
+
+            def peek(self, size=0) -> bytes:
+                if self.loop_count > 0:
+                    self.loop_count -= 1
+                    return b""
+                return peek_return
+
+            def read(self, _size: int = -1) -> bytes:
+                return self.contents
+
+        class MockPopen(subprocess.Popen):
+            def __init__(self, *_args, **_kwargs):
+                # No effect, other than to quiet the linter
+                None if True else super().__init__([])
+                if popen_fail:
+                    raise ValueError(
+                        "MockPopen pretending it was called with invalid arguments"
+                    )
+                self.stdout = MockBufferedReader(stdout_contents)
+                self.stderr = MockBufferedReader(stderr_contents)
+                self.returncode = None
+                self.loop_count = wait_cnt
+
+            def poll(self) -> Optional[int]:
+                if self.loop_count > 0:
+                    self.loop_count -= 1
+                    return None
+                self.returncode = poll_return
+                return poll_return
+
+            def kill(self) -> None:
+                raise AssertionError(
+                    "This test doesn't expect the stream to be closed."
+                )
```

Comment:
Hmm.

Reply:
Well, now...since I added a call to `kill()` in `Tarball.extract()` for the creation-timeout, now this has to be a "real mock"...so, let's have another round!